Closes #1800 - Adding Symbol Table Overview documentation (#1873)
Bears-R-Us/arkouda: training/SYMBOL_TABLE_OVERVIEW.md

# Overview

The purpose of this document is to serve as a high-level guide to the structure and elements within Arkouda's Symbol Table. This guide does not cover creating or refactoring new Symbol Table Entry types. 

<a id="toc"></a>
## Table of Contents
1. [Symbol Table Overview](#overview)
2. [Symbol Table Structure](#symStruct)
3. [Creating a New Entry From an Array](#addSym)
4. [Finding an Entry in the Symbol Table](#findSym)
5. [Accessing Symbol Table Entry Data](#accessSym)

<a id="overview"></a>
## Symbol Table Overview

> The Symbol Table is Arkouda's "database" that tracks the components every Arkouda object is composed of. At a fundamental level it functions similar to a relational database, however the design and access methods are quite different.

Reviewer: I second this because I have no clue what a relational database and that would make me feel like it was a complex object that I couldn't understand. I think lookup table feels more approachable

> The Symbol Table is Arkouda's "database" that tracks the components every Arkouda object is composed of. At a fundamental level it functions similar to a relational database, however the design and access methods are quite different.

Author: I reworded this to use `lookup table` and mention that it uses a Chapel `Map`

<a id="symStruct"></a>
## Symbol Table Structure

> The Symbol Table stores `Entries`. The Entry types are designed in a simple hierarchical structure as follows:

Author: Updated to your suggestion

- `AbstractSymEntry`
  - `GenSymEntry`
    - `SymEntry`
    - `SegArraySymEntry`
    - `SegStringSymEntry`
  - `CompositeSymEntry`

>     - `GroupBySymEntry`

Author: I agree that the `enum` object isn't necessary for this document

> The root-level entry type is `AbstractSymEntry`. All other entry types inherit from `AbstractSymEntry` or one of its ancestors. `AbstractSymEntry` does not store any data, only metadata information, such as name and entry type.

Reviewer: I was thinking rewording it to `All other entry types inherit from AbstractSymEntry either directly or through one of its children`, but I'm not sure if that's too wordy

> The root-level entry type is `AbstractSymEntry`. All other entry types inherit from `AbstractSymEntry` or one of its ancestors. `AbstractSymEntry` does not store any data, only metadata information, such as name and entry type.

Author: Reworded this with a mix of both suggestions

> `GenSymEntry` is the first type directly inheriting from `AbstractSymEntry`. `GenSymEntry` is used to define the main metadata necessary for entry types that contain a single set of values of a single data type.

Author: I reworded this based on your comment, hopefully it makes more sense now

`SymEntry` is the entry type used for basic one-dimensional arrays of numeric values and their metadata. This is the most commonly used entry type.

`SegArraySymEntry` tracks the component pieces of a `SegmentedArray` object. These components are all instances of `SymEntry` types.

`SegStringSymEntry`tracks the component pieces of a `SegmentedStrings` object. The values of a `SegStringSymEntry` are stored in a `uint8` `SymEntry`. These components are all instances of `SymEntry` types.

> `CompositeSymEntry` is used to indicate that child `Entries` inheriting from this entry type contain multiple `data-adding` objects. This means that types under `CompositeSymEntry` contain multiple objects of varying types. For example, `SegArraySymEntry` contains three `SymEntry` entries, but all three are of a single type and relate to a single SegArray Object. `GroupBySymEntry` contains five `SymEntry` entries, but each `SymEntry` could be of a different type from the others, i.e. `keyNamesEntry` would be of type `str` while `permEntry` would be of type `int64`.

Reviewer: also they aren't all of the same type right? the `values` has a different type (can be anything) than `offsets` (has to be int)

> `CompositeSymEntry` is used to indicate that child `Entries` inheriting from this entry type contain multiple `data-adding` objects. This means that types under `CompositeSymEntry` contain multiple objects of varying types. For example, `SegArraySymEntry` contains three `SymEntry` entries, but all three are of a single type and relate to a single SegArray Object. `GroupBySymEntry` contains five `SymEntry` entries, but each `SymEntry` could be of a different type from the others, i.e. `keyNamesEntry` would be of type `str` while `permEntry` would be of type `int64`.

Author: `SegArraySymEntry` does contain three `SymEntry` values. It used to be two until `lengths` was added

> `CompositeSymEntry` is used to indicate that child `Entries` inheriting from this entry type contain multiple `data-adding` objects. This means that types under `CompositeSymEntry` contain multiple objects of varying types. For example, `SegArraySymEntry` contains three `SymEntry` entries, but all three are of a single type and relate to a single SegArray Object. `GroupBySymEntry` contains five `SymEntry` entries, but each `SymEntry` could be of a different type from the others, i.e. `keyNamesEntry` would be of type `str` while `permEntry` would be of type `int64`.

Reviewer 2: @joshmarshall1 - you are right... I was referencing the old code. My fault.

> `GroupBySymEntry` is used to track component pieces of a `GroupBy` object. It accomplishes this task by tracking five separate `SymEntry` instances that each relate to one of the components.

Author: Updated

<a id="addSym"></a>
## Creating a New Entry From an Array

> If your Chapel module creates an array of data that you then want to either persist or return a reference to back to the Client, then you would need to add that array to the symbol table. This is pretty simple and only requires a single line of code to accomplish.

Author: Added

```chapel
  // Assume we have a borrowed Symbol Table (SymTab) called 'st'
  
  // Array Creation
  var myArray: [0..#4] int = [1, 2, 3, 4, 5];
  
  // Get the next available name in the symbol table and the add the entry using that name
  var newName = st.nextName();
  st.addEntry(newName, new shared SymEntry(myArray));
```

Once added, that array can then be accessed from the symbol table using the value of `newName`.

<a id="findSym"></a>
## Finding an Entry in the Symbol Table

There are two ways of finding if a given string matches the name of a Symbol Table entry:

```chapel
  st.contains(string); // Returns bool
  st.findAll(regex string); // Returns [] string
```

`st.contains(string)` is used to verify the existence of an entry that matches the provided string. This function will return `True` if it exists within the Symbol Table and `False` if it does not.

> `st.findAll(regex string)` is used when you have a partial name that can map to multiple entries in a set. An example of when this is used is when `Categorical` objects are registered to the symbol table. These objects register 4 separate entries, one for each piece of the `Categorical` needed to rebuild it on the client. 

Author: I added a blurb for `checkTable` and highlighted that it is the most commonly used. I'm not sure how I missed this the first time around

> If you have a set of data with entries `dataset_entry_0`, `dataset_entry_1`, and `dataset_entry_2`, you could manually find each entry if you know how many entries are in your dataset. However, this function is most useful for when you may not know exactly how many entries you have. By passing the portion of the name these three entries have in common, `dataset_entry_`, followed by regex notation for "Any number, 1 or more times", `/d+`. We can then get a list of all the matching names in the symbol table and operate on all the dataset pieces.

Author: We don't need to know the number of entries ahead of time. I reworded this so it's hopefully clearer

```chapel
  var entryList = st.findAll("dataset_entry_/d+");
  // entryList: [`dataset_entry_0`, `dataset_entry_1`, 'dataset_entry_2']
```

You can then loop through the returned array of names to access the data and perform operations on each entry in the set.

<a id="accessSym"></a>
## Accessing Symbol Table Entry Data

Depending on your functionality requirements, there are a couple different ways to access a symbol table entry.

### Passing the Entry to the Client

> If you need to return an entry to the python for computations, the `SymTab.attrib(name)` method will return the entry's attributes, which are required by the arkouda client to reference the entry and use the underlying data. You can then use the return message to `attach` to the entry and access the data as you would with any other Python object.

Reviewer: `to the python` -> `to python`

Author: Fixed

> If you need to return an entry to the python for computations, the `SymTab.attrib(name)` method will return the entry's attributes, which are required by the arkouda client to reference the entry and use the underlying data. You can then use the return message to `attach` to the entry and access the data as you would with any other Python object.

Author: Updated to highlight this point

```chapel
  var values = entry.a;
  var domain = entry.aD;
```

Each entry type has its own set of properties that can be accessed using the `.` notation, similar to accessing class properties in Python objects. Below is a list of all accessible properties for each entry type within the hierarchy:

- `AbstractSymEntry`
  - `entryType`
  - `assignableTypes`
  - `name`
  - `GenSymEntry`
    - All properties of `AbstractSymEntry`
    - `dtype`
    - `itemsize`
    - `size`
    - `ndim`
    - `shape`
    - `SymEntry`
      - All properties of `GenSymEntry`
      - `etype` - Chapel's version of `dtype`
      - `a` - Stores the values of the array
      - `aD` - Stores the domain information
    - `SegArraySymEntry`
      - All properties of `GenSymEntry`
      - `segmentsEntry`
      - `valuesEntry`
      - `lengthsEntry`
    - `SegStringSymEntry`
      - All properties of `GenSymEntry`
      - `offsetsEntry`
      - `bytesEntry`
  - `CompositeSymEntry`
    - All properties of `AbstractSymEntry`
    - `ndim`
    - `size`
    - `GroupBySymEntry`
      - All properties of `CompositeSymEntry`
      - `keyNamesEntry`
      - `keyTypesEntry`
      - `segmentsEntry`
      - `permEntry`

>       - `ukIndEntry`

Reviewer 2: I would break this out instead of also having the entries tiered:

- AbstractSymEntry
    - Properties listed
- GenSymEntry
    - Properties listed
etc.

I do like how you reference the properties of the parent.

Reviewer: I agree, the current way makes it hard to tell where the properties end and the child class begins once rendered
- `AbstractSymEntry`
  - `entryType`
  - `assignableTypes`
  - `name`
  - `GenSymEntry`
    - All properties of `AbstractSymEntry`
    - `dtype`
    - `itemsize`
    - `size`
    - `ndim`
    - `shape`
    - `SymEntry`
      - All properties of `GenSymEntry`

Author: Reformatted to be on one level with spacing. I think it's clearer now, but you can be the judge

# Overview

The purpose of this document is to serve as a high-level guide to the structure and elements within Arkouda's Symbol Table. This guide does not cover creating or refactoring new Symbol Table Entry types. 

For more information regarding the creation of new Symbol Table Entry types, view the [Symbol Table Entry Creation](https://github.com/Bears-R-Us/arkouda/blob/master/training/SYMBOL_TABLE_ENTRY_CREATION.md) document.

<a id="toc"></a>
## Table of Contents
1. [Symbol Table Overview](#overview)
2. [Symbol Table Structure](#symStruct)
3. [Creating a New Entry From an Array](#addSym)
4. [Finding an Entry in the Symbol Table](#findSym)
5. [Accessing Symbol Table Entry Data](#accessSym)

<a id="overview"></a>
## Symbol Table Overview

> The Symbol Table is Arkouda's "database" that tracks the components every Arkouda object is composed of. It serves as a `lookup table` that uses a Chapel `Map` which allows Arkouda to persist objects until the server is shutdown or the object is deleted.

Reviewer 2: ```suggestion
The Symbol Table is implemented using a Chapel Map and serves as a `lookup table` that uses a Chapel `Map` which allows Arkouda to persist objects until the server is shutdown or the object is deleted.

```

I think we need to totally remove any reference to a database

# Overview

The purpose of this document is to serve as a high-level guide to the structure and elements within Arkouda's Symbol Table. This guide does not cover creating or refactoring new Symbol Table Entry types. 

For more information regarding the creation of new Symbol Table Entry types, view the [Symbol Table Entry Creation](https://github.com/Bears-R-Us/arkouda/blob/master/training/SYMBOL_TABLE_ENTRY_CREATION.md) document.

<a id="toc"></a>
## Table of Contents
1. [Symbol Table Overview](#overview)
2. [Symbol Table Structure](#symStruct)
3. [Creating a New Entry From an Array](#addSym)
4. [Finding an Entry in the Symbol Table](#findSym)
5. [Accessing Symbol Table Entry Data](#accessSym)

<a id="overview"></a>
## Symbol Table Overview

> The Symbol Table is implemented using a Chapel Map and serves as a `lookup table` that uses a Chapel `Map` which allows Arkouda to persist objects until the server is shutdown or the object is deleted.

Reviewer: ```suggestion
The Symbol Table is implemented using a Chapel Map and serves as a `lookup table` which allows Arkouda to persist objects until the server is shutdown or the object is deleted.

```

Reviewer 2: I think @pierce314159 suggestion makes sense here to avoid repeating that a Chapel Map is used.

Author: Updated to use this suggestion

  
The root-level entry type is `AbstractSymEntry`. All other entry types inherit from `AbstractSymEntry` directly or from one of its children. `AbstractSymEntry` does not store any data, only metadata information, such as name and entry type.

`GenSymEntry` is the first type directly inheriting from `AbstractSymEntry`. `GenSymEntry` is used to define the main metadata necessary for entry types that contain a single set of values and its necessary components. For example, a `SegArraySymEntry` stores three `SymEntries`, `values`, `segments`, and `lengths`, however these three pieces are all necessary to build a single instance of a `SegArray` object and reference a single set of data values.

`SymEntry` is the entry type used for basic one-dimensional arrays of numeric values and their metadata. This is the most commonly used entry type.

`SegArraySymEntry` tracks the component pieces of a `SegmentedArray` object. These components are all instances of `SymEntry` types.

`SegStringSymEntry`tracks the component pieces of a `SegmentedStrings` object. The values of a `SegStringSymEntry` are stored in a `uint8` `SymEntry`. These components are all instances of `SymEntry` types.

> `CompositeSymEntry` is used to indicate that child `Entries` inheriting from this entry type contain multiple `data-adding` objects. This means that types under `CompositeSymEntry` contain multiple objects of varying types, which may not all be needed to conceptualize the Object. For example, `GroupBySymEntry` contains five `SymEntry` entries. One of these, `keyNamesEntry` could contain `n` number of references to additional data values. These could be of multiple types, and it is possible to build a GroupBy object with all or only part of the stored values.

Reviewer: `This means that types under CompositeSymEntry contain multiple objects of varying types`

I get what this is saying but I feel like it could be confusing since types has two meanings here. Maybe we could reword it to something like

`This means that entry types under CompositeSymEntry contain multiple objects of varying types`

Also I feel like `Object` doesn't need to be capitalized here

Author: Updated

```

You can then loop through the returned array of names to access the data and perform operations on each entry in the set.

<a id="accessSym"></a>
## Accessing Symbol Table Entry Data

Depending on your functionality requirements, there are a couple different ways to access a symbol table entry.

### Passing the Entry to the Client

If you need to return an entry to python, the `SymTab.attrib(name)` method will return the entry's metadata, which is required by the arkouda client to reference the entry when operations are sent to the server to be performed on the object. You can then use the return message to `attach` to the entry and view the data as you would with any other Python object.

It is important to note that on the client after `attaching` to a Symbol Table Entry, the client does not store the data or perform computations on the data. The client is only storing the metadata and any computations on the object are performed on the server.

More information on passing data between the client and server can be found [here](https://github.com/Bears-R-Us/arkouda/blob/master/training/MESSAGING_OVERVIEW.md).

### Working with Entries in Chapel

> When working with a basic `SymEntry` entry type, you can easily access the data values within the entry by accessing the property `a`. To access the domain information, access the property `aD`. For example, assume we have a SymTab entry variable called `entry`.

Reviewer: maybe drop `information` here, since we're getting the domain itself rather than info about it

Reviewer 2: You could also say `To access the domain field` or `To access the domain object`. Not sure if this is necessary or if @pierce314159's suggestion to just remove information is sufficient.

Author: Dropped `information`. I think this is sufficient to just drop information

```chapel
  var values = entry.a;

>   var domain = entry.aD;

Reviewer: we should name our `domain` variable since that is a reserved keyword in chapel

Author: Renamed to `entryDomain`
